[REEF-1378] Creating Evaluator Manager for Fault Tolerant

jwang98052 opened the pull request:
Adding Evaluator Manager which manages allocated evaluators and failed evaluators mainly for book keeping, tracking failed evaluators and master evaluator.
Adding test cases for it

JIRA: [REEF-1378](https://issues.apache.org/jira/browse/REEF-1378)
This closes

afchung commented:
I'll do a pass.

afchung commented on lang/cs/Org.Apache.REEF.IMRU.Tests/TestEvaluatorManager.cs:
```diff
@@ -0,0 +1,187 @@
+        /// Test case: no master Evaluator is requested
+        /// </summary>
+        [Fact]
+        public void TestNoMasterEvalautor()
```
Evaluator

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,311 @@
+        /// <param name="evaluatorId"></param>
+        internal void RecordFailedEvaluator(string evaluatorId)
+        {
+            RomoveAllocatedEvaluator(evaluatorId);
```
Remove

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,311 @@
+        /// <returns></returns>
+        internal bool IsMasterEvaluatorFailed
+        {
+            get { return _failedEvaluators.Any(eid => IsMasterEvaluatorId(eid)); }
```
`Contains`.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,311 @@
+        /// <summary>
+        /// Reset failed Evaluator collection
+        /// </summary>
+        internal void ResetFailedEvalutors()
```
Evaluators

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU.Tests/TestEvaluatorManager.cs:
```diff
@@ -0,0 +1,187 @@
+        [Fact]
+        public void TestRemoveInvalidEvaluators()
+        {
+            var evalutorManager = CreateTestEvaluators(3, 1);
```
Rename all `evalutorManager` to `evaluatorManager`.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorSpecification.cs:
```diff
@@ -0,0 +1,46 @@
+        /// <summary>
+        /// Size of the memory for the Evaluator request
+        /// </summary>
+        internal int Megabytes { get; set; }
```
Can the `set` be `private`?

jwang98052 replied:
yes

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorSpecification.cs:
```diff
@@ -0,0 +1,46 @@
+        /// <summary>
+        /// Number of core for the Evaluator request
+        /// </summary>
+        internal int Core { get; set; }
```
Can the `set` be `private`?

jwang98052 replied:
yes

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,311 @@
+        /// <returns></returns>
+        internal bool IsMasterEvaluator(IAllocatedEvaluator evaluator)
+        {
+            return evaluator.EvaluatorBatchId.Equals(MasterBatchId);
```
Why not call `IsMasterEvaluatorId`?

afchung replied:
I see, because it hasn't been set yet. Can you rename to `IsMasterEvaluatorBatch`?

jwang98052 replied:
Renamed to IsEvaluatorForMaster. We need to check if the allocated evaluator is for master.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,311 @@
+        /// <param name="evaluatorRequestor"></param>
+        /// <param name="updateEvaluatorSpecification"></param>
+        /// <param name="mapperEvaluatorSpecification"></param>
+        internal EvaluatorManager(
```
Will this later on be injected?

jwang98052 replied:
Yes, we can do that.

markusweimer replied:
This strikes me as a bit too dynamic to easily make statically configured. Maybe leave this off tang for now. We can revisit over time.

afchung commented on lang/cs/Org.Apache.REEF.IMRU.Tests/TestEvaluatorManager.cs:
```diff
@@ -0,0 +1,187 @@
+        /// Test case: request two master Evaluators
+        /// </summary>
+        [Fact]
+        public void TestTwoMasterEvalautor()
```
Rename all instances of `Evalautor` to `Evaluator`.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,311 @@
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on anResetMasterEvaluatorId
```
Remove `ResetMasterEvaluatorId`

jwang98052 replied:
Yes :)

afchung commented:
@jwang98052 done with a first pass. Mostly fine, but a few concerns:

- Relying on the outside to reset state is potentially dangerous. If the caller forgets, then we will be in an inconsistent state. Example:
```
evaluatorManager.ResetMasterEvaluatorId();

// Will have one extra, because MasterEvaluatorId has been reset!
evaluatorManager.NumberOfFailedMappers();
```
- Non-thread safe properties can be *very* tricky to debug.

jwang98052 commented:
@afchung I have addressed your comments.

markusweimer commented:
I'm doing a pass.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,317 @@
+        internal const string MasterBatchId = "MasterBatchId";
+        internal const string MapperBatchId = "MapperBatchId";
+
+        private readonly ISet<string> _allocatedEvaluators = new HashSet<string>();
```
maybe rename these to `allocatedEvaluatorIDs`, as we don't store `AllocatedEvaluator` instances.

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,317 @@
+        internal const string MapperBatchId = "MapperBatchId";
+
+        private readonly ISet<string> _allocatedEvaluators = new HashSet<string>();
+        private readonly ISet<string> _failedEvaluators = new HashSet<string>();
```
Same, consider renaming to `_failedEvaluatorIDs`

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,317 @@
+        /// <summary>
+        /// Checks if all the expected Evaluators are allocated.
+        /// </summary>
+        internal bool AreAllEvaluatorsAllocated
```
I have a C# style question on this one: This isn't a property, but actually computes something. Is it OK to use properties for that? I'm asking because @afchung recently hit an issue where computed properties gave him trouble in combination with LINQ.

I'm OK with this being a property because it is side-effect free. Is that the right rule?

afchung replied:
This can get tricky after multiple threads enter the picture, and you may need to end up locking every property access.

jwang98052 replied:
One of the principle of using property vs. method is "They should not have dependencies on each other.". Looks like it is safer to use method for those that have computing which has dependency on other properties. I will change those unless it only returns a value.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorSpecification.cs:
```diff
@@ -0,0 +1,46 @@
+    /// <summary>
+    /// Specification for an Evaluator
+    /// </summary>
+    internal sealed class EvaluatorSpecification
```
Can't we use the evaluator request class for this?

jwang98052 replied:
EvaluatorRequest set number in the constructor. We want to leave the number later when calling the request methods. Those request methods can be called again and again during the recovery with the same specification.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,317 @@
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (NumberOfAllocatedEvaluators == _totalExpectedEvaluators)
```
`==` --> `>=`

jwang98052 replied:
done

markusweimer commented:
I've done a pass. Looks good to me overall. With minor comments. Will leave some design thoughts over on the JIRA.

afchung commented:
@jwang98052 taking a second pass now.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,317 @@
+                SetMasterEvaluatorId(evaluator.Id);
+            }
+
+            if (_masterEvaluatorId == null && NumberOfAllocatedEvaluators + 1 == _totalExpectedEvaluators)
```
You can move the logic to after `_allocatedEvaluators.Add`.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU.Tests/TestEvaluatorManager.cs:
```diff
@@ -0,0 +1,187 @@
+            evalutorManager.RecordFailedEvaluator(EvaluatorIdPrefix + 2);
+            Assert.Equal(2, evalutorManager.NumberOfMissingEvaluators);
+            evalutorManager.ResetFailedEvaluators();
+            Assert.Equal(0, evalutorManager.NumberofFailedMappers);
```
Perhaps more properties can be validated here?

jwang98052 replied:
done

afchung commented:
@jwang98052 Looks good to me. Only a few more minor comments. Thanks!

jwang98052 commented:
@markusweimer @afchung I have addressed your review comments.

markusweimer commented:
LGTM. +1 for merging.

dkm2110 commented:
Doing a pass now

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs:
```diff
@@ -0,0 +1,312 @@
+                    .SetMegabytes(_mapperEvaluatorSpecification.Megabytes)
+                    .SetNumber(numEvaluators)
+                    .SetCores(_mapperEvaluatorSpecification.Core)
+                    .SetEvaluatorBatchId(MapperBatchId)
```
This is a problem here. Suppose earlier you asked for 100 mappers with batch id as MapperBatchId. Now if 5 fail and you use the function again to request for them, you cannot use the same batch Id. It will throw an error.

@afchung @markusweimer  Can you confirm this logic?

jwang98052 replied:
Just re-submitted one with the same batch id, it successes. This is in local runtime. Not sure about Yarn.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/TestEvaluatorManager.cs:
```diff
@@ -0,0 +1,189 @@
+        }
+
+        /// <summary>
+        /// Create a an EvaluatorManager and add mock Evaluators for testing
```
Remove "a"

jwang98052 replied:
done

dkm2110 commented:
@jwang98052  pointed one potential bug.

jwang98052 commented:
@dkm2110 I have addressed your review comments. 
My test shows batch id can be reused within one driver. But I still changed it to append a sequence number, that would be also good for log info.

dkm2110 commented:
LGTM

afchung commented:
Looks good to me, I'll test and merge.
